### PR TITLE
Add getter/parser for PVGIS hourly-radiation

### LINKED ISSUE #849: getter/parser for PVGIS hourly-radiation
**Is your feature request related to a problem? Please describe.**
Related to #845 

**Describe the solution you'd like**
Similar to `get_pvgis_tmy` retrieve pvgis hourly radiation data from their api

**Describe alternatives you've considered**
Pvgis is becoming a popular resource more and more people are asking me for it, it is nice because it is a global collection of several different radiation databases including nsrdb and others, and different from cams, the data is complete, ie it has air temperature, wind speed as well as all 3 components of irradiance

**Additional context**
This would be part of the `iotool` sub-package. There's already a `pvgis.py` module with a getter for tmy be data

### PR DESCRIPTION
**Pull request by AdamRJensen**
- [x] Closes #849
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

Add getter/parser for PVGIS hourly-radiation

### CODE REVIEW

**Comment by AdamRJensen** [comment]
@mikofski @wholmgren Should the tests for the hourly radiation functions just mimick those for the TMY functions?

**Comment by mikofski** [comment]
That's probably okay. I think the goals for testing should be,  "have I sent the request correctly?"  "Is the response what I expected?" and "did it get parsed right?" Make sure to mark api calls with `pytest.remotedata` & `flaky` using the timeout constants from `conftest`. Thanks.  PS i'm sure my tests could be improved. Don't feel constrained to repeat my mistakes ;)

**Comment by AdamRJensen** [comment]
@kanderso-nrel  @mikofski 
The pvgis hourly API is able to return data in three different formats (specified by the outputformat keyword): json, csv, and basic.

The basic format is essentially the same as the csv option, but without a header or metadata. Is there any interest in supporting this option?

Note that for PVGIS TMY the basic format has a header, which it is NOT supposed to have according to the development team behind PVGIS.

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:261]: Can the `.replace('\n', '').replace('\r', '')` pattern used here and elsewhere be replaced with `.strip()`?

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:123]: I think this is only ever a dict, when can it be a list?

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:294]: `map_variables` isn't in the docstring `Parameters` list

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:299]: Having the `read_` function be able to take a buffer goes against the `get/read/parse` pattern we sort of follow, but I might like this way better.  I'm curious what other people think about it.

File: pvlib/tests/iotools/test_pvgis.py
Comment (kanderso-nrel) [inline pvlib/tests/iotools/test_pvgis.py:154]: Would it be cleaner to define fixtures for the expected dataframes?  pytest can't parametrize directly across fixtures (unfortunately), but it has a `getfixturevalue` function that can be used as a workaround, for example: https://stackoverflow.com/a/64348247/1641381

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:123]: Nope, you are right. It was an artefact copied from get_pvgis_tmy where the csv option is returning metadata as a list.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:299]: I copied this functionality from the existing `read_pvgis_tmy`

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:33]: The description from the metadata file says "Reflected irradiance on the inclined plane (plane of the array)". But I think that it is fair to assume that it's `poa_sky_diffuse`

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:261]: Nice solution. Seems to work fine.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:81]: At least I was consistent! But I have fixed it now.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:111]: Yes it should. I have raised an issue, see #1250.

File: docs/sphinx/source/whatsnew/v0.9.0.rst
Comment (AdamRJensen) [reply docs/sphinx/source/whatsnew/v0.9.0.rst]: I changed it to: 

Added :func:`~pvlib.iotools.read_pvgis_hourly` and
:func:`~pvlib.iotools.get_pvgis_hourly` for reading and retrieving hourly
solar radiation data and PV power output from PVGIS. (:pull:`1186`,
:issue:`849`)

Is that clearer?

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [reply pvlib/iotools/pvgis.py:298]: Yeah, like how `pd.read_csv` can take a filepath or a buffer; you don't have to switch to `pd.parse_csv` for when you want to pass a file-like object instead of a filename.  Also less clutter in the API listings with the "two-in-one" approach.  Thoughts from @wholmgren or anyone else?

Comment (wholmgren) [reply pvlib/iotools/pvgis.py:298]: I agree that it's a nicer interface for users. I am less excited about the gymnastics we have to do to support it. The `try/except AttributeError` pattern below feels lacking for some reason. Maybe I'd like it more if we used `hasattr` - it's more explicit, but it does go against the python pattern of preferring try/except to checks. That would also let you avoid repeated calls to the pvlib helper function. In any case, how do other libraries handle this?

File: pvlib/tests/iotools/test_pvgis.py
Comment (AdamRJensen) [reply pvlib/tests/iotools/test_pvgis.py:137]: The solution has been implemented and did indeed make the tests more readable and compact.

Comment (AdamRJensen) [reply pvlib/tests/iotools/test_pvgis.py:137]: The solution has been implemented and did indeed make the tests more readable and compact.

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [reply pvlib/iotools/pvgis.py:294]: Without thinking too much about it, one clean approach might be a decorator that does something like (a much simpler version of) pandas's `_get_filepath_or_buffer`: https://github.com/pandas-dev/pandas/blob/master/pandas/io/common.py#L226

File: pvlib/tests/iotools/test_pvgis.py
Comment (kanderso-nrel) [inline pvlib/tests/iotools/test_pvgis.py:277]: Does the output here need to be checked?  Or is this just to verify that no error is raised?

**Review by kanderso-nrel — COMMENTED** [review]
Looking good @AdamRJensen, some minor comments below.  I think this is ready for a detailed look from someone else now, maybe @mikofski for familiarity with the original PVGIS functions?  Though obviously anyone is welcome to review.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:139]: Yes, good catch!

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:213]: Yes indeed. You have a keen eye!

File: pvlib/tests/iotools/test_pvgis.py
Comment (AdamRJensen) [reply pvlib/tests/iotools/test_pvgis.py:277]: I'm not sure I understand the question?

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:194]: I do like the one liner, nice and clean. From what I've read, isinstance is often preferable to type as it supports inheritance.

File: pvlib/tests/iotools/test_pvgis.py
Comment (AdamRJensen) [reply pvlib/tests/iotools/test_pvgis.py:277]: The purpose of this test function is solely to check that the input parameters not tested in the previous test are formatted correctly in the request.

The output would not change significantly, hence I do not seen any additional benefit in adding a third test file.

Comment (kanderso-nrel) [reply pvlib/tests/iotools/test_pvgis.py:277]: Thanks, I forgot that requests-mock will raise an error if you try to fetch an unexpected URL.  Makes sense.

File: pvlib/iotools/pvgis.py
Comment (wholmgren) [inline pvlib/iotools/pvgis.py:255]: How do we feel about the possibility of an infinite loop if `time,` isn't in the response? Could happen if people are using local files that have been edited or if the API changes. Seems that we should at least break if `line` is empty (end of file). Guessing that'll cause an exception somewhere down the line but could add one here (I don't think it's all that important).

File: pvlib/iotools/pvgis.py
Comment (wholmgren) [inline pvlib/iotools/pvgis.py:266]: to be defensive (as above) I suggest adding `line == ''` too

File: pvlib/iotools/pvgis.py
Comment (wholmgren) [inline pvlib/iotools/pvgis.py:274]: I try to avoid `inplace=True` because pandas is so inconsistent about it. I think Wes McKinney once said it was a bad idea. In any case, it's more characters than `data = `!

File: pvlib/tests/iotools/test_pvgis.py
Comment (wholmgren) [inline pvlib/tests/iotools/test_pvgis.py:236]:
```suggestion
    # Open local test file containing McClear monthly data
```

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:63]: Not relevant = not used, so yes ignored. Any preference in terminology here?

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:59]: I'll implement this along with your other suggestions in a new commit.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:192]: If you were to pass a float right now it would fail because line 190 would attempt to do start.year which throw an error. 

I'm not quite sure what you are suggesting? Are you suggesting also allowing floats, e.g.: 

    params['startyear'] = start if (isinstance(start, int) | isinstance(start, float)) else start.year

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:268]: Perhaps a better alternative would be

    if line.strip() == '':

From a quick investigation it seems this would catch all three cases.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:276]: Thank you for the explanation, I just keep learning! :)

Consider it changed.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: Very good point! I think initially I had stuck to the order of the variables in the documentation. I have now changed it such that start and end come right after latitude and longitude, which is similar to other get functions.

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [reply pvlib/iotools/pvgis.py:75]: I don't recall seeing it elsewhere either, but sphinx doesn't seem to care so neither do I

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [reply pvlib/iotools/pvgis.py:76]: Just FYI that language comes from the PVGIS docs: https://ec.europa.eu/jrc/en/PVGIS/docs/noninteractive

Comment (wholmgren) [reply pvlib/iotools/pvgis.py:76]: Minor preference for "ignored"

File: pvlib/iotools/pvgis.py
Comment (wholmgren) [reply pvlib/iotools/pvgis.py:194]: There's not a clean 1-liner for what I was suggesting, so let's ignore. Sorry for the noise. FYI `isinstance` can take a tuple of types e.g. `isinstance(start, (int, float))`.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: I think this is a good idea. Wouldn't it be better to raise an error instead of breaking though (as it would mean that no data section was detected)?

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: I've replaced all occurences of `.replace('\n', '')` and `.replace('\r\n', '')` with `.strip()`. This has surely made the code a lot cleaner.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:223]: This check is necessary as outputformat='basic' is an option recognized by PVGIS but one that we have chosen not to support.

The returning of None was copied from `_parse_pvgis_tmy` - can remove it from there too?

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:75]: The [Numpy Styleguide](https://numpydoc.readthedocs.io/en/latest/format.html) allows a space, `=`, or `:`.

I'll leave the `:` unless yvall have any other objections.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:76]: Ok. I'm changing it to ignored.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [inline pvlib/iotools/pvgis.py:113]: @kanderso-nrel Would it be preferred here to also have the URL as a module variable instead of an input?

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [inline pvlib/iotools/pvgis.py:220]: @wholmgren @kanderso-nrel Given that `read_pvgis_hourly` is able to take a file buffer, couldn't we just replace all these lines with:

    return read_pvgis_hourly(io.StringIO(res.text), pvgis_format=outputformat,
                             map_variables=map_variables)

File: pvlib/iotools/pvgis.py
Comment (mikofski) [reply pvlib/iotools/pvgis.py:289]: I'm not crazy about decorators b/c IMHO the abstraction can make the code hard to read and maintain. However, if it's a pattern that we can reuse a lot, then I do think it's worth the effort.

I prefer EAFP try-except vs type-checking, which I think is more Pythonic b/c an object might be able to do the action even if it isn't a subclass of any of the types in the check, but as Emerson says, "A foolish consistency..." so if there's a simpler, cleaner way than try-except, then I'll be in favor.
>Simple is better than complex.
>Complex is better than complicated.

The reason I broke with the get/read/parse pattern for PVGIS was because it had the 4 types, and the number of parsing functions seemed more complicated then just combining read & parse together.

A pattern I usually like for this is recursion. Not sure why I didn't use it here, something like:

```python
def read_fbuf_or_filename(f):
    try:
        f.seek(0)  # or some other method only fbuf has
    except AttributeError:
        # str and Path don't have read
        with pathlib.Path(f).open() as g:
            # now that we have a buffer use recursion to parse
            return read_fbuf_or_filename(g)
    return parse_fbuf(f)
```

I think the reason I didn't like this here is because I wanted to use pandas `to_blah` and the EPW methods which work on file buffers?

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [reply pvlib/iotools/pvgis.py:113]: Oh, maybe.  Pro: it would shorten a long parameter list.  Con: inconsistent with the PVGIS TMY function.  If you do remove it and make it a module variable, it should be named such that there is room to eventually make the PVGIS TMY URL a module variable as well.

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [reply pvlib/iotools/pvgis.py:220]: makes sense to me, no sense in duplicating the `outputformat` -> `_parse_pvgis_hourly_*` mapping.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:245]: @wholmgren How do you feel about raising an IOError?

        elif line == '':  # If end of file is reached
            raise IOError('No data section was detected. File has probably '
                          'been modified since being downloaded from PVGIS')

Comment (wholmgren) [reply pvlib/iotools/pvgis.py:245]: I'm all for raising an exception. I'm not sure about `IOError`. It was [merged into `OSError`](https://docs.python.org/3/library/exceptions.html#OSError) and I think there's discussion of deprecating it in python 4+. I suggest `ValueError` since it relates to a problem with the values within the file/buffer.

File: pvlib/iotools/pvgis.py
Comment (wholmgren) [reply pvlib/iotools/pvgis.py]: Thanks for explaining. So `ValueError('invalid outputformat')` seems appropriate here.

Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: Now that read_bsrn is called from the get_bsrn function, this error is instead handled by read_bsrn which checks that outputformat is either 'csv' or 'json'.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:113]: PVGIS hourly and TMY share the same base URL, so the change would have to be made to both at the same time. I'm ok for leaving it for now though

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:136]: I don't recognize this `!!!` syntax -- is it valid RST?  The docs aren't rendering correctly, maybe because of this line.  I'd have expected to see something like this:

```
    .. warning:: PVGIS provides access to a number of different
        solar radiation...
```

https://pvlib-python--1186.org.readthedocs.build/en/1186/generated/pvlib.iotools.get_pvgis_hourly.html

Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: @kanderso-nrel I have added a section called "Hint" instead. The PR is now complete and can be merged if there are no other comments.

Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: @kanderso-nrel I have added a section called "Hint" instead. The PR is now complete and can be merged if there are no other comments.

Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: @kanderso-nrel I have added a section called "Hint" instead. The PR is now complete and can be merged if there are no other comments.

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:200]: Can this logic for `start` be condensed the same way it was for `end`?  I.e. merge the `elif` block into the preceding `if`.

File: pvlib/iotools/pvgis.py
Comment (kanderso-nrel) [inline pvlib/iotools/pvgis.py:192]: maybe incomplete comment

**Review by kanderso-nrel — APPROVED** [review]
LGTM thanks for seeing this through @AdamRJensen!  Two very minor points below, then I promise I'll hit the merge button.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py]: It was left from a previously deleted comment.

File: pvlib/iotools/pvgis.py
Comment (AdamRJensen) [reply pvlib/iotools/pvgis.py:198]: No good reason for this, I guess I had just forgotten. It is fixed now and ready to merge! :D

**Pull request merged**
